Managed gateway groundwork

johanstokking opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

#### Summary

<!--
A short summary, referencing related issues:
References #0000, etc.
Pull requests may close issues (using Closes #0000) only for minor changes not needing tests on a staging environment.
Typically, issues should be closed manually after validation.
-->

References https://github.com/TheThingsIndustries/lorawan-stack/issues/4256
Groundwork for https://github.com/TheThingsNetwork/lorawan-stack/pull/7178

This adds groundwork for managed gateways in The Things Stack.

#### Changes

<!-- What are the changes made in this pull request? -->

- Add groundwork for TTI gateway protocol as GS frontend
- Add The Things Gateway Controller client and integration in Gateway Configuration Server
- Add events multiplexer to add additional event streams, in this case managed gateway events that come from The Things Gateway Controller
- Add missing checks for gateway tokens
- Resolve some technical debt, including but not limited to
   - Existing WebSocket listeners in GS
   - Lint warnings
   - Obsolete settings
   - Structure of GCS
   - Missing GS IO test cases and values (suppressed for UDP and LBS LNS that have limited support for this)

#### Testing

Testing of this PR is mostly through unit tests and tedious manual tests with WebSockets and TTS API calls.

Testing steps for QA sign off will be part of https://github.com/TheThingsNetwork/lorawan-stack/pull/7178.

#### Notes for Reviewers

<!--
NOTE: This section is optional.

Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->

@KrishnaIyer is the main reviewer.

This is draft until:

- [x] Testing section is filled out
- [x] GS frontend is implemented and tested
- [x] More integration testing is done
- [x] Commit trail is clean
- [x] Infrastructure PR is (almost) ready
- [x] Changelog is updated
- [x] CI pass

#### Checklist

<!-- Make sure that this pull request is complete. -->

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [ ] Testing: The steps/process to test this feature are clearly explained including testing for regressions.
- [x] Infrastructure: If infrastructural changes (e.g., new RPC, configuration) are needed, a separate issue is created in the infrastructural repositories.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

johanstokking committed:
all: Add TLS auth info to peer if available

johanstokking committed:
all: Return client certificate from peer

johanstokking committed:
gs: Simplify Semtech web socket listeners

johanstokking committed:
gs: Add The Things Industries gateway frontend

johanstokking committed:
gs: Remove obsolete setting

johanstokking committed:
gs: Harmonize config description

johanstokking committed:
gs: Resolve lint warnings

johanstokking committed:
gs: Provide HTTP client

johanstokking committed:
gs: Fix setting absolute time in gateway status to server time

This field is used to convey the gateway's absolute time, not the
server's time.

johanstokking committed:
gs: Return authenticated context

johanstokking committed:
gs: Support customizing the component config

johanstokking committed:
gs: Resolve lint warnings

johanstokking committed:
api: Define managed gateway

johanstokking committed:
api: Define services to manage WiFi and Ethernet profiles

johanstokking committed:
dev: Add go.thethings.industries/pkg dependency

johanstokking committed:
gcs: Rename package for The Things Kickstarter Gateway

johanstokking committed:
all: Regenerate messages

johanstokking committed:
all: Add Gateway Controller config and client

johanstokking committed:
gcs: Add gateway managed via Gateway Controller

johanstokking committed:
all: Implement events subscription multiplexer

johanstokking committed:
all: Respect context cancellation in stream partition iteration

johanstokking committed:
all: Add IEEE mappings groundwork

johanstokking committed:
cli: Configure TTGC as event stream

johanstokking committed:
gcs: Add managed gateway event subscriber client

johanstokking committed:
console: Make gateway event filter more explicit

johanstokking committed:
gcs: Skip first managed gateway update

johanstokking committed:
gcs: Handle unmanaged gateways gracefully

KrishnaIyer commented on api/ttn/lorawan/v3/gateway.proto:
```diff
@@ -249,6 +251,272 @@ message Gateways {
+    select: true,
+    set: true
+  };
+  string profile_id = 1 [(validate.rules).string = {
```
Should this be a first class identifier (similar to `AlertProfileIdentifiers`)?

johanstokking replied:
Hmm I think it's something opaque for TTS, it's an external ID, not a TTS concept. I'm a bit divided to do this, because then we'll have to define messages for potentially many external IDs.

johanstokking replied:
Coming back on this, let's keep this a string as it is a TTS-external identifier.

KrishnaIyer reviewed (commented):
Reviewed API, GCS and the events mux. Looks good to me.

johanstokking committed:
gs: Support gateway tokens in frontend test

johanstokking committed:
is: Authenticate gateway tokens

johanstokking committed:
is,gcs: Support gateway tokens in mock IS

johanstokking committed:
all: Expose encoding gateway tokens to string

johanstokking committed:
gs: Fix potential panic

This happens when the TTL is shorter than the update interval. This
typically only happens during tests with a debugger attached, but may
also happen with bad config.

johanstokking committed:
gs: Improve logging

johanstokking committed:
gs: Treat dropping CRC and invalid LoRaWAN frames differently

johanstokking committed:
gs: Improve test logging

johanstokking committed:
gs: Test Tx acknowledgment error

johanstokking committed:
gs: Fix frequencies based on frequency plan

johanstokking committed:
gs: Set signal RSSI for LoRa data rates

johanstokking committed:
gs: Fix event data of Tx failure

johanstokking committed:
gcs: Implement managing ethernet profiles

johanstokking committed:
gcs,api: Expose shared setting for managed gateway WiFi and ethernet profiles

johanstokking committed:
gs: Fix connecting unregistered gateways

johanstokking committed:
gcs: Shorten event descriptions

johanstokking committed:
gs: Return authenticated context

johanstokking committed:
gs: Move bandwidth to Tx channel

johanstokking committed:
dev: Regenerate messages

johanstokking committed:
dev: Skip .pb.go files for headers checks

johanstokking committed:
all: Trust The Things Industries Root CA for Gateway Controller connections

The Gateway Controller is only available as hosted service operated by
The Things Industries.

johanstokking committed:
all: Allow passing additional root CAs via context

johanstokking committed:
gs: Use port 1889

Port 1888 is used by a proprietary frontend.

johanstokking committed:
gs: Trust The Things Industries Root CA

johanstokking committed:
api: Fix typo

johanstokking committed:
all: Swallow mux stream non-retryable errors and retry others

johanstokking committed:
gs: Enable customizing comparing Rx metadata

johanstokking committed:
gs: Clear signal RSSI as UDP protocol V1 does not support it

johanstokking committed:
gs: Enable skipping Tx acknowledgment test when it's not successful

johanstokking committed:
gs: Clear signal RSSI and skip non-successful Tx acknowledgment test

johanstokking committed:
all: Set namespaces to avoid race conditions with parallel tests

johanstokking committed:
dev: Update changelog

johanstokking committed:
dev: Tidy dependencies

johanstokking committed:
api: Format files

KrishnaIyer commented on pkg/ttgc/config.go:
```diff
@@ -0,0 +1,29 @@
+	Enabled     bool                 `name:"enabled" description:"Enable The Things Gateway Controller"`
+	GatewayEUIs []types.EUI64Prefix  `name:"gateway-euis" description:"Gateway EUI prefixes that are managed by The Things Gateway Controller"` //nolint:lll
+	Address     string               `name:"address" description:"The address of The Things Gateway Controller"`
+	Domain      string               `name:"domain" description:"The domain of this cluster"`
```
Can't this be fetched from the global configuration?

johanstokking replied:
How? We have public addresses in component configs because we don't have a global config for the domain.

It would be helpful but that is beyond the scope of this PR. Unless you have a reference to a global config key that we can use instead here.

KrishnaIyer reviewed (approved):
Great stuff!

johanstokking committed:
all: Fix panic

johanstokking committed:
all: Suppress NotFound non-retryable stream errors
